src/Sound.ts

```diff
@@ -174,44 +192,158 @@ export default class Sound {
+// nodes were completely disconnected from the chain or otherwise had never
+// been applied. This allows for clean discarding of effect nodes when returned
+// to the initial value.
+//
+// The order of this array matches AudioEngine's effects list in scratch-audio.
+// Earlier in the list is closer to the EffectChain input node; later is closer
+// to its target (output). Note that a non-"patch" effect's position in the
+// array has no bearing on effect behavior, since it isn't part of the chain
+// system.
+//
+// Note that this descriptor list is fairly easy to build on, if we'd like to
+// add more audio effects in the future. (Scratch used to have more, but they
+// were removed - see commit ff6cd4a - because they depended on an external
+// library and were too processor-intensive to support on some devices.)
+const PanEffect: EffectDescriptor<
+  true,
+  "pan",
+  { leftGain: GainNode; rightGain: GainNode }
+> = {
+  name: "pan",
+  initial: 0,
+  minimum: -100,
+  maximum: 100,
+  isPatch: true,
+  makeNodes() {
+    const aCtx = Sound.audioContext;
+    const input = aCtx.createGain();
+    const leftGain = aCtx.createGain();
+    const rightGain = aCtx.createGain();
+    const channelMerger = aCtx.createChannelMerger(2);
+    const output = channelMerger;
+    input.connect(leftGain);
+    input.connect(rightGain);
+    leftGain.connect(channelMerger, 0, 0);
+    rightGain.connect(channelMerger, 0, 1);
+    return { input, output, leftGain, rightGain, channelMerger };
+  },
+  set(value, { leftGain, rightGain }) {
+    const p = (value + 100) / 200;
+    const leftVal = Math.cos((p * Math.PI) / 2);
+    const rightVal = Math.sin((p * Math.PI) / 2);
+    const { currentTime } = Sound.audioContext;
+    const { decayWait, decayDuration } = EffectChain;
+    leftGain.gain.setTargetAtTime(
+      leftVal,
+      currentTime + decayWait,
+      decayDuration
+    );
+    rightGain.gain.setTargetAtTime(
+      rightVal,
+      currentTime + decayWait,
+      decayDuration
+    );
+  },
+} as const;
+const PitchEffect: EffectDescriptor<false, "pitch", never> = {
+  name: "pitch",
+  initial: 0,
+  isPatch: false,
+  set(value, sound) {
+    const interval = value / 10;
+    const ratio = Math.pow(2, interval / 12);
+    sound.setPlaybackRate(ratio);
+  },
+} as const;
+const VolumeEffect: EffectDescriptor<true, "volume", { node: GainNode }> = {
+  name: "volume",
+  initial: 100,
+  minimum: 0,
+  maximum: 100,
+  resetOnStart: false,
+  resetOnClone: true,
+  isPatch: true,
+  makeNodes() {
+    const node = Sound.audioContext.createGain();
+    return {
+      input: node,
+      output: node,
+      node,
+    };
+  },
+  set(value, { node }) {
+    node.gain.linearRampToValueAtTime(
+      value / 100,
+      Sound.audioContext.currentTime + EffectChain.decayDuration
+    );
+  },
+} as const;
+
+const effectDescriptors = [PanEffect, PitchEffect, VolumeEffect] as const;
```

Comment:
It's allowing us to get the complete list of effect descriptors at compile-time. It allows it to type as `[PatchDescriptor<...>, PatchlessDescriptor<...>, PatchDescriptor<...>]` with each specific type parameter filled in completely. Without the `as const`, it would type as an array of a union of all three.